enhancements/monitoring/alerting-consistency.md

```diff
@@ -0,0 +1,195 @@
+(dataplane disruption)
+* inability to start any new pod other than capacity.  EG, if the API server
+were to restart or need to be rescheduled, it would not able to start on another
+host
+
+In otherwords, critical alerts are something that require someone to get out
+of bed in the middle of the night and fix something **right now** or they will
+be faced with a disaster.
+
+An example of something that is NOT a critical alert:
+* MCO and/or related components are completely dead/crash looping.
+
+Even though the above is quite significant, there is no risk of immediate loss
+of the cluster.
+
+You might be thinking to yourself: "But my component is **super** important, and
+if it's not working, the user can't do X, which user definitely will want."  All
+that is probably true, but that doesn't make your alert "critical" it just makes
+it worthy of an alert.
+
+### Warning Alerts
+
+TL/DR: fix this soon, or some things won't work and upgrades will probably be
+blocked.
+
+If your alert does not meet the criteria in "Critical Alerts" above, it belongs
+to the warning level or lower.
+
+Most alerts probably fit in this category.  This level should represent degraded
+or unavailable services that do not jeopardize the immediate life of cluster.
+
+Timeline:  ~60 minutes
+
+60 minutes?  That seems high!  That's because it is high.  We want to reduce
+the noise.  We've done a lot to make clusters and operators auto-heal
+themselves.  The whole idea is that if a condition has persisted for more than
+60 minutes, it's unlikely to be resolved without intervention any time later.
+
+#### Example 1: All machine-config-daemons are crashlooping.
+
+That seems significant
+and it probably is, but nobody has to get out of bed to fix this **right now**.
+But, what if a machine dies, I won't be able to get a replacement?  Yeah, that
+is probably true, but how often does that happen?  Also, that's an entirely
+separate set of concerns.
+
+#### Example 2:  I have an operator that needs at least 3 of something to work.
+* With 2/3 replicas, warning alert after 60M.
+* With 1/3 replicas, warning alert after 10M.
+* With 0/3 replicas, warning alert after 5M.
+
+Q: How long should we wait until the above conditions rises to a 'critical'
+alert?
+
+A: It should never rise to the level of a critical alert, it's not critical.  If
+it was, this section would not apply.
+
+#### Example 3:  I have an operator that only needs 1 replica to function
+
+Well, if the cluster can upgrade with only 1 replica, and the the service is
+not unavailable despite other replicas being unavailable, this can probably
+be just an info-level alert.  We should have a "minor" alert, TBD.  But for now,
+"info" is the closest thing.
+
+#### Q: What if a particular condition will block upgrades?
+
+A: It's a warning level alert.
+
+
+### Documentation Required
+
+Every alert needs to have a URL to relevant documentation to resolve the issue.
```

Comment:
There is a RFE for this already, but it is not possible to do this right now, due to disconnected evns from last time I spoke with monitoring PM. This is essentially runbooks no?

Reply:
I don't know what disconnected has to do with having a URL?  Even if the cluster is in an isolated environment, that doesn't mean the person looking at the alerts is.  At the worst case, someone can just copy and paste the URL.

Reply:
This was the issue at the time, it also was a problem due to a lot of alerts coming from upstream mixins, which means URLs would be different. There should be a ticket for runbooks on the monitoring board with more details on this, its being worked on or at least last I heard. Best to contact Christian our PM for more details here.

Reply:
We shouldn't use upstream alert mixins as part of the core alerting.  How many alerts are we even talking about?  Can't be more than a few dozen, probably some of which we don't want or need.  We need to trim down the superfluous stuff.

Reply:
All alerts for prometheus, alertmanager, prometheus-operator, kubernetes, and majority for k8s api, k8s scheduler, k8s controller manager. All in all, around 150 alerts and recording rules. The majority of alerts in OpenShift **IS** coming from upstream and this is aligned with "upstream first" mentality.

We should use alerts from upstream due to the same reason we use kubernetes as base. We shouldn't invent everything ourselves.

Reply:
IMO, this is reactive and not proactive.  There's a reason most projects in upstream are alpha/beta statuses.  We should selective choose and groom the alerts we care about to deliver the best experience to our users.  If we're just bulk-importing the alerting rules, it means we (probably) don't have resolution docs to go with them.

Reply:
#298 has me poking around in this space for conditions, since we have more control over there.  I like the link idea, and they don't have to be external.  We can have alert messages link to local, console-hosted docs or UXes that help unpack the issue being reported in more context, like [this][1].  And paragraph-or-two level of discussion on likely impact and suggested recovery procedures, I don't see why we couldn't either grow out the alert message, or attach a new field/annotation to alerts to carry that information, even for alerts defined in upstream mixins.

[1]: https://github.com/openshift/cluster-version-operator/blob/3640003453ad3186ce9ce360e4788bcfbacbd5e0/install/0000_90_cluster-version-operator_02_servicemonitor.yaml#L51-L53

Reply:
> IMO, this is reactive and not proactive. 

All alerts I mentioned are created upstream in projects largely managed by OpenShift engineers. Most of them are tweaked because of OpenShift. I don't see how this is reactive.

Reply:
Just because something exists upstream doesn't mean it's good and doesn't mean it belongs downstream.  We should bring in the quality pieces and leave the rest.

Here's an example: "KubePodNotReady"

That is a warning level alert.  I'm assuming that is coming from upstream.  Is having this alert helping many people, or is it contributing to the noise part of signal-to-noise?

A better alert would be for the specific application (user created alert) or for a trend, if the trend of non-ready pods is increasing over time.

Reply:
In the SRE team orthogonal effort to reduce alert noise (which is ongoing for many months, predates this PR, yadda yadda), we will be forced to improve the upstream alerts.  That is a circular workflow that we worked with the monitoring team on.  SRE team knows what to do when something like this comes up.